rflx/specification/parser.py

```diff
@@ -1,358 +1,981 @@
+import rflx.expression as expr
+import rflx.model as model
+import rflx.statement as stmt
+from rflx.error import Location, RecordFluxError, Severity, Subsystem, fail
+from rflx.identifier import ID, StrID
+from rflx.specification.const import RESERVED_WORDS
+
 from .cache import Cache
 
 log = logging.getLogger(__name__)
 
 
-class Parser:
-    def __init__(self, skip_verification: bool = False, cached: bool = False) -> None:
-        self.skip_verification = skip_verification
-        self.__specifications: OrderedDict[Path, Optional[Specification]] = OrderedDict()
-        self.__evaluated_specifications: Set[ID] = set()
-        self.__types: List[Type] = [*BUILTIN_TYPES.values(), *INTERNAL_TYPES.values()]
-        self.__sessions: List[Session] = []
-        self.__cache = Cache(cached)
+def node_location(node: RFLXNode, filename: Path = None) -> Location:
+    start = node.token_start.sloc_range
+    end = node.token_end.sloc_range
+    return Location(
+        start=(start.start.line, start.start.column),
+        source=Path(str(filename) if filename else "<stdin>"),
+        end=(end.end.line, end.end.column),
+    )
 
-    def parse(self, *specfiles: Path) -> None:
-        error = RecordFluxError()
 
-        for f in specfiles:
-            error.extend(self.__determine_dependencies(f))
+def diagnostics_to_error(
+    diagnostics: List[Diagnostic], error: RecordFluxError, specfile: Path = None
+) -> bool:
+    """
+    Append langkit diagnostics to RecordFlux error. Return True if error occured.
+    """
 
-        with Pool() as p:
-            for f, s, e in p.map(
-                self._parse, [f for f, s in self.__specifications.items() if s is None]
-            ):
-                self.__specifications[f] = s
-                error.extend(e)
+    if len(diagnostics) == 0:
+        return False
 
-            p.close()
-            p.join()
+    for diag in diagnostics:
+        loc = diag.sloc_range
+        error.append(
+            diag.message,
+            Subsystem.PARSER,
+            Severity.ERROR,
+            Location(
+                start=(loc.start.line, loc.start.column),
+                source=specfile,
+                end=(loc.end.line, loc.end.column),
+            ),
+        )
+    return True
 
-        error.propagate()
 
-    def __determine_dependencies(
-        self, specfile: Path, transitions: List[Tuple[str, ID]] = None
-    ) -> RecordFluxError:
-        if specfile in self.__specifications:
-            self.__specifications.move_to_end(specfile)
-        else:
-            self.__specifications[specfile] = None
+def create_description(description: Description = None) -> Optional[str]:
+    if description:
+        return description.text.split('"')[1]
+    return None
 
-        if not transitions:
-            transitions = []
 
-        error = RecordFluxError()
+def create_transition(
+    transition: Transition, filename: Path = None, package: ID = None
+) -> model.Transition:
```

Comment:
Many of these `create_*` functions have an optional `package` parameter, although it is not required. I removed some of them and still all tests pass:

```diff
diff --git a/rflx/specification/parser.py b/rflx/specification/parser.py
index 610c9731..7246dd2b 100644
--- a/rflx/specification/parser.py
+++ b/rflx/specification/parser.py
@@ -92,37 +92,35 @@ def create_description(description: Description = None) -> Optional[str]:
     return None
 
 
-def create_transition(
-    transition: Transition, filename: Path = None, package: ID = None
-) -> model.Transition:
+def create_transition(transition: Transition, filename: Path = None) -> model.Transition:
     if transition.kind_name not in ("Transition", "ConditionalTransition"):
         raise NotImplementedError(f"Transition kind {transition.kind_name} unsupported")
     target = create_id(transition.f_target, filename)
     condition: expr.Expr = expr.TRUE
     description = create_description(transition.f_description)
     if transition.kind_name == "ConditionalTransition":
-        condition = create_bool_expression(transition.f_condition, filename, package)
+        condition = create_bool_expression(transition.f_condition, filename)
     return model.Transition(target, condition, description, node_location(transition, filename))
 
 
 def create_reset(
-    reset: Statement, filename: Path = None, _package: ID = None, location: Location = None
+    reset: Statement, filename: Path = None, location: Location = None
 ) -> stmt.Statement:
     return stmt.Reset(create_id(reset.f_identifier, filename), location=location)
 
 
 def create_assignment(
-    assignment: Statement, filename: Path = None, package: ID = None, location: Location = None
+    assignment: Statement, filename: Path = None, location: Location = None
 ) -> stmt.Statement:
     return stmt.Assignment(
         create_id(assignment.f_identifier, filename),
-        create_expression(assignment.f_expression, filename, package),
+        create_expression(assignment.f_expression, filename),
         location,
     )
 
 
 def create_list_attribute(
-    expression: Statement, filename: Path = None, package: ID = None, location: Location = None
+    expression: Statement, filename: Path = None, location: Location = None
 ) -> stmt.Statement:
     attrs = {
         "Append": stmt.Append,
@@ -134,24 +132,20 @@ def create_list_attribute(
 
     return constructor(
         create_id(expression.f_identifier, filename),
-        create_expression(expression.f_expression, filename, package),
+        create_expression(expression.f_expression, filename),
         location=location,
     )
 
 
-def create_statement(
-    statement: Statement, filename: Path = None, package: ID = None
-) -> stmt.Statement:
+def create_statement(statement: Statement, filename: Path = None) -> stmt.Statement:
     handlers: Dict[
-        str, Callable[[Statement, Optional[Path], Optional[ID], Optional[Location]], stmt.Statement]
+        str, Callable[[Statement, Optional[Path], Optional[Location]], stmt.Statement]
     ] = {
         "Reset": create_reset,
         "Assignment": create_assignment,
         "ListAttribute": create_list_attribute,
     }
-    return handlers[statement.kind_name](
-        statement, filename, package, node_location(statement, filename)
-    )
+    return handlers[statement.kind_name](statement, filename, node_location(statement, filename))
 
 
 def create_state(state: State, filename: Path = None) -> model.State:
@@ -523,18 +517,16 @@ def create_binding(
 
 
 def create_variable_decl(
-    declaration: VariableDecl, filename: Path = None, package: ID = None, location: Location = None
+    declaration: VariableDecl, filename: Path = None, location: Location = None
 ) -> decl.BasicDeclaration:
     initializer = (
-        create_expression(declaration.f_initializer, filename, package)
+        create_expression(declaration.f_initializer, filename)
         if declaration.f_initializer
         else None
     )
     return decl.VariableDeclaration(
         create_id(declaration.f_identifier, filename),
-        model.qualified_type_identifier(
-            create_id(declaration.f_type_identifier, filename), package
-        ),
+        model.qualified_type_identifier(create_id(declaration.f_type_identifier, filename)),
         initializer,
         location=location,
     )
@@ -575,15 +567,13 @@ def create_channel_decl(
 
 
 def create_renaming_decl(
-    declaration: RenamingDecl, filename: Path = None, package: ID = None, location: Location = None
+    declaration: RenamingDecl, filename: Path = None, location: Location = None
 ) -> decl.BasicDeclaration:
-    selected = create_expression(declaration.f_expression, filename, package)
+    selected = create_expression(declaration.f_expression, filename, None)
     assert isinstance(selected, expr.Selected)
     return decl.RenamingDeclaration(
         create_id(declaration.f_identifier, filename),
-        model.qualified_type_identifier(
-            create_id(declaration.f_type_identifier, filename), package
-        ),
+        model.qualified_type_identifier(create_id(declaration.f_type_identifier, filename), None),
         selected,
         location,
     )
@@ -718,15 +708,13 @@ def create_expression(expression: Expr, filename: Path = None, package: ID = Non
     return EXPRESSION_MAP[expression.kind_name](expression, filename, package, location)
 
 
-def create_declaration(
-    declaration: Expr, filename: Path = None, package: ID = None
-) -> decl.BasicDeclaration:
+def create_declaration(declaration: Expr, filename: Path = None) -> decl.BasicDeclaration:
     location = node_location(declaration, filename)
     handlers: Dict[str, Callable[..., decl.BasicDeclaration]] = {
         "VariableDecl": create_variable_decl,
         "RenamingDecl": create_renaming_decl,
     }
-    return handlers[declaration.kind_name](declaration, filename, package, location)
+    return handlers[declaration.kind_name](declaration, filename, location)
 
 
 def create_formal_declaration(
```

IMHO all useless function parameters should be removed.

Reply:
Removed in dd63f7df, thanks for the patch!

Reply:
I think there are more instances where the optional `package` parameter is not required. Could you please check that?

Reply:
`create_message_aspects` still had one which I removed in 82f746924afc718e8409f1fa8caef331ad6707f8. The other functions with unused `package` parameters are called from a dict (e.g. `EXPRESSION_MAP`) and thus required the same signature as other functions (where that parameter is used).

Reply:
I was able to remove all optional `package` parameters. I even removed a part in `create_variable` explicitly checking if `package` is set. All these changes had no effect on any test. Either all this code is unnecessary or tests are missing.

```diff
diff --git a/rflx/specification/parser.py b/rflx/specification/parser.py
index df44cccb..2872f558 100644
--- a/rflx/specification/parser.py
+++ b/rflx/specification/parser.py
@@ -280,7 +280,7 @@ def create_array(
 
 
 def create_numeric_literal(
-    expression: Expr, _filename: Path = None, _package: ID = None, location: Location = None
+    expression: Expr, _filename: Path = None, location: Location = None
 ) -> expr.Expr:
     num = expression.text.split("#")
     if len(num) == 1:
@@ -300,24 +300,24 @@ OPERATIONS: Dict[str, Type[expr.BinExpr]] = {
 
 
 def create_binop(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
     loc = node_location(expression, filename)
     if expression.f_op.kind_name in OPERATIONS:
         return OPERATIONS[expression.f_op.kind_name](
-            create_expression(expression.f_left, filename, package),
-            create_expression(expression.f_right, filename, package),
+            create_expression(expression.f_left, filename),
+            create_expression(expression.f_right, filename),
             location=loc,
         )
     if expression.f_op.kind_name in BOOLEAN_OPERATIONS:
         return BOOLEAN_OPERATIONS[expression.f_op.kind_name](
-            create_expression(expression.f_left, filename, package),
-            create_expression(expression.f_right, filename, package),
+            create_expression(expression.f_left, filename),
+            create_expression(expression.f_right, filename),
             location=loc,
         )
 
-    left = create_math_expression(expression.f_left, filename, package)
-    right = create_math_expression(expression.f_right, filename, package)
+    left = create_math_expression(expression.f_left, filename)
+    right = create_math_expression(expression.f_right, filename)
     if expression.f_op.kind_name in MATH_OPERATIONS:
         return MATH_OPERATIONS[expression.f_op.kind_name](left, right, location=loc)
     if expression.f_op.kind_name in MATH_COMPARISONS:
@@ -337,12 +337,12 @@ MATH_OPERATIONS = {
 
 
 def create_math_binop(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
     if expression.f_op.kind_name in MATH_OPERATIONS:
         return MATH_OPERATIONS[expression.f_op.kind_name](
-            create_math_expression(expression.f_left, filename, package),
-            create_math_expression(expression.f_right, filename, package),
+            create_math_expression(expression.f_left, filename),
+            create_math_expression(expression.f_right, filename),
             location=node_location(expression, filename),
         )
     raise NotImplementedError(
@@ -364,24 +364,24 @@ BOOLEAN_OPERATIONS = {
 
 
 def create_bool_binop(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
     if expression.f_op.kind_name in MATH_COMPARISONS:
         return MATH_COMPARISONS[expression.f_op.kind_name](
-            create_math_expression(expression.f_left, filename, package),
-            create_math_expression(expression.f_right, filename, package),
+            create_math_expression(expression.f_left, filename),
+            create_math_expression(expression.f_right, filename),
             location=node_location(expression, filename),
         )
     if expression.f_op.kind_name in BOOLEAN_OPERATIONS:
         return BOOLEAN_OPERATIONS[expression.f_op.kind_name](
-            create_bool_expression(expression.f_left, filename, package),
-            create_bool_expression(expression.f_right, filename, package),
+            create_bool_expression(expression.f_left, filename),
+            create_bool_expression(expression.f_right, filename),
             location=node_location(expression, filename),
         )
     if expression.f_op.kind_name in OPERATIONS:
         return OPERATIONS[expression.f_op.kind_name](
-            create_expression(expression.f_left, filename, package),
-            create_expression(expression.f_right, filename, package),
+            create_expression(expression.f_left, filename),
+            create_expression(expression.f_right, filename),
             location=node_location(expression, filename),
         )
     raise NotImplementedError(
@@ -390,40 +390,35 @@ def create_bool_binop(
 
 
 def create_paren_bool_expression(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
-    return create_bool_expression(expression.f_data, filename, package)
+    return create_bool_expression(expression.f_data, filename)
 
 
 def create_paren_math_expression(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
-    return create_math_expression(expression.f_data, filename, package)
+    return create_math_expression(expression.f_data, filename)
 
 
 def create_paren_expression(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
-    return create_expression(expression.f_data, filename, package)
+    return create_expression(expression.f_data, filename)
 
 
 def create_variable(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     if expression.f_identifier.text.lower() in ("true", "false"):
         return expr.Variable(create_id(expression.f_identifier), location=location)
-    if package:
-        return expr.Variable(
-            model.qualified_type_identifier(create_id(expression.f_identifier, filename), package),
-            location=location,
-        )
     return expr.Variable(create_id(expression.f_identifier, filename), location=location)
 
 
 def create_math_attribute(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
-    inner = create_expression(expression.f_expression, filename, package)
+    inner = create_expression(expression.f_expression, filename)
     if expression.f_kind.kind_name == "AttrLast":
         return expr.Last(inner)
     if expression.f_kind.kind_name == "AttrFirst":
@@ -436,9 +431,9 @@ def create_math_attribute(
 
 
 def create_attribute(
-    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+    expression: Expr, filename: Path = None, _location: Location = None
 ) -> expr.Expr:
-    inner = create_expression(expression.f_expression, filename, package)
+    inner = create_expression(expression.f_expression, filename)
     if expression.f_kind.kind_name == "AttrLast":
         return expr.Last(inner)
     if expression.f_kind.kind_name == "AttrFirst":
@@ -461,16 +456,16 @@ def create_attribute(
 
 
 def create_array_aggregate(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     return expr.Aggregate(
-        *[create_math_expression(v, filename, package) for v in expression.f_values],
+        *[create_math_expression(v, filename) for v in expression.f_values],
         location=location,
     )
 
 
 def create_string_literal(
-    expression: Expr, _filename: Path = None, _package: ID = None, location: Location = None
+    expression: Expr, _filename: Path = None, location: Location = None
 ) -> expr.Expr:
     return expr.String(
         expression.text.split('"')[1],
@@ -479,21 +474,21 @@ def create_string_literal(
 
 
 def create_call(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     return expr.Call(
         create_id(expression.f_identifier, filename),
-        [create_expression(a, filename, package) for a in expression.f_arguments],
+        [create_expression(a, filename) for a in expression.f_arguments],
         location=location,
     )
 
 
 def create_quantified_expression(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     param_id = create_id(expression.f_parameter_identifier, filename)
-    iterable = create_expression(expression.f_iterable, filename, package)
-    predicate = create_expression(expression.f_predicate, filename, package)
+    iterable = create_expression(expression.f_iterable, filename)
+    predicate = create_expression(expression.f_predicate, filename)
     if expression.f_operation.kind_name == "QuantifierAll":
         return expr.ForAllIn(param_id, iterable, predicate, location)
     if expression.f_operation.kind_name == "QuantifierSome":
@@ -503,14 +498,14 @@ def create_quantified_expression(
 
 
 def create_binding(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     bindings: Mapping[Union[str, ID], expr.Expr] = {
-        create_id(b.f_identifier, filename): create_expression(b.f_expression, filename, package)
+        create_id(b.f_identifier, filename): create_expression(b.f_expression, filename)
         for b in expression.f_bindings
     }
     return expr.Binding(
-        create_expression(expression.f_expression, filename, package), bindings, location
+        create_expression(expression.f_expression, filename), bindings, location
     )
 
 
@@ -533,7 +528,6 @@ def create_variable_decl(
 def create_private_type_decl(
     declaration: FormalPrivateTypeDecl,
     filename: Path = None,
-    _package: ID = None,
     location: Location = None,
 ) -> decl.FormalDeclaration:
     return decl.TypeDeclaration(
@@ -544,7 +538,6 @@ def create_private_type_decl(
 def create_channel_decl(
     declaration: FormalChannelDecl,
     filename: Path = None,
-    _package: ID = None,
     location: Location = None,
 ) -> decl.FormalDeclaration:
     readable = False
@@ -567,11 +560,11 @@ def create_channel_decl(
 def create_renaming_decl(
     declaration: RenamingDecl, filename: Path = None, location: Location = None
 ) -> decl.BasicDeclaration:
-    selected = create_expression(declaration.f_expression, filename, None)
+    selected = create_expression(declaration.f_expression, filename)
     assert isinstance(selected, expr.Selected)
     return decl.RenamingDeclaration(
         create_id(declaration.f_identifier, filename),
-        model.qualified_type_identifier(create_id(declaration.f_type_identifier, filename), None),
+        model.qualified_type_identifier(create_id(declaration.f_type_identifier, filename)),
         selected,
         location,
     )
@@ -580,7 +573,6 @@ def create_renaming_decl(
 def create_function_decl(
     declaration: FormalFunctionDecl,
     filename: Path = None,
-    package: ID = None,
     location: Location = None,
 ) -> decl.FormalDeclaration:
     arguments = []
@@ -590,7 +582,7 @@ def create_function_decl(
                 decl.Argument(
                     create_id(p.f_identifier, filename),
                     model.qualified_type_identifier(
-                        create_id(p.f_type_identifier, filename), package
+                        create_id(p.f_type_identifier, filename)
                     ),
                 )
             )
@@ -603,62 +595,62 @@ def create_function_decl(
 
 
 def create_negation(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
-    expression = create_math_expression(expression.f_data, filename, package)
+    expression = create_math_expression(expression.f_data, filename)
     assert isinstance(expression, expr.Number)
     return expr.Number(-expression.value, expression.base, location)
 
 
 def create_concatenation(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
-    left = create_expression(expression.f_left, filename, package)
-    right = create_expression(expression.f_right, filename, package)
+    left = create_expression(expression.f_left, filename)
+    right = create_expression(expression.f_right, filename)
     assert isinstance(left, expr.Aggregate)
     assert isinstance(right, expr.Aggregate)
     return expr.Aggregate(*(left.elements + right.elements), location=location)
 
 
 def create_comprehension(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     condition = (
-        create_bool_expression(expression.f_condition, filename, package)
+        create_bool_expression(expression.f_condition, filename)
         if expression.f_condition
         else expr.TRUE
     )
     return expr.Comprehension(
         create_id(expression.f_iterator, filename),
-        create_expression(expression.f_array, filename, package),
-        create_expression(expression.f_selector, filename, package),
+        create_expression(expression.f_array, filename),
+        create_expression(expression.f_selector, filename),
         condition,
         location,
     )
 
 
 def create_selected(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     return expr.Selected(
-        create_expression(expression.f_expression, filename, package),
+        create_expression(expression.f_expression, filename),
         create_id(expression.f_selector, filename),
         location=location,
     )
 
 
 def create_conversion(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     return expr.Conversion(
         create_id(expression.f_target_identifier, filename),
-        create_expression(expression.f_argument, filename, package),
+        create_expression(expression.f_argument, filename),
         location=location,
     )
 
 
 def create_message_aggregate(
-    expression: Expr, filename: Path = None, package: ID = None, location: Location = None
+    expression: Expr, filename: Path = None, location: Location = None
 ) -> expr.Expr:
     values: Mapping[StrID, expr.Expr] = {}
     if expression.f_values.kind_name == "NullMessageAggregate":
@@ -666,7 +658,7 @@ def create_message_aggregate(
     elif expression.f_values.kind_name == "MessageAggregateAssociations":
         values = {
             create_id(c.f_identifier, filename): create_expression(
-                c.f_expression, filename, package
+                c.f_expression, filename
             )
             for c in expression.f_values.f_associations
         }
@@ -700,10 +692,10 @@ EXPRESSION_MAP = {
 }
 
 
-def create_expression(expression: Expr, filename: Path = None, package: ID = None) -> expr.Expr:
+def create_expression(expression: Expr, filename: Path = None) -> expr.Expr:
 
     location = node_location(expression, filename)
-    return EXPRESSION_MAP[expression.kind_name](expression, filename, package, location)
+    return EXPRESSION_MAP[expression.kind_name](expression, filename, location)
 
 
 def create_declaration(declaration: Expr, filename: Path = None) -> decl.BasicDeclaration:
@@ -716,7 +708,7 @@ def create_declaration(declaration: Expr, filename: Path = None) -> decl.BasicDe
 
 
 def create_formal_declaration(
-    declaration: Expr, filename: Path = None, package: ID = None
+    declaration: Expr, filename: Path = None
 ) -> decl.FormalDeclaration:
     location = node_location(declaration, filename)
     handlers: Dict[str, Callable[..., decl.FormalDeclaration]] = {
@@ -724,11 +716,11 @@ def create_formal_declaration(
         "FormalFunctionDecl": create_function_decl,
         "FormalPrivateTypeDecl": create_private_type_decl,
     }
-    return handlers[declaration.kind_name](declaration, filename, package, location)
+    return handlers[declaration.kind_name](declaration, filename, location)
 
 
 def create_math_expression(
-    expression: Expr, filename: Path = None, package: ID = None
+    expression: Expr, filename: Path = None
 ) -> expr.Expr:
     location = node_location(expression, filename)
     handlers = {
@@ -743,11 +735,11 @@ def create_math_expression(
         "SelectNode": create_selected,
         "ArrayAggregate": create_array_aggregate,
     }
-    return handlers[expression.kind_name](expression, filename, package, location)
+    return handlers[expression.kind_name](expression, filename, location)
 
 
 def create_bool_expression(
-    expression: Expr, filename: Path = None, package: ID = None
+    expression: Expr, filename: Path = None
 ) -> expr.Expr:
     location = node_location(expression, filename)
     handlers = {
@@ -760,7 +752,7 @@ def create_bool_expression(
         "Binding": create_binding,
         "SelectNode": create_selected,
     }
-    return handlers[expression.kind_name](expression, filename, package, location)
+    return handlers[expression.kind_name](expression, filename, location)
 
 
 def create_modular(
@@ -773,7 +765,7 @@ def create_modular(
 ) -> model.ModularInteger:
     return model.ModularInteger(
         identifier,
-        create_math_expression(modular.f_mod, filename, identifier.parent),
+        create_math_expression(modular.f_mod, filename),
         type_location(identifier, modular),
     )
 
@@ -793,11 +785,11 @@ def create_range(
             Severity.ERROR,
             node_location(rangetype, filename),
         )
-    size = create_math_expression(rangetype.f_size.f_value, filename, identifier.parent)
+    size = create_math_expression(rangetype.f_size.f_value, filename)
     return model.RangeInteger(
         identifier,
-        create_math_expression(rangetype.f_first, filename, identifier.parent),
-        create_math_expression(rangetype.f_last, filename, identifier.parent),
+        create_math_expression(rangetype.f_first, filename),
+        create_math_expression(rangetype.f_last, filename),
         size,
         type_location(identifier, rangetype),
     )
@@ -1096,7 +1088,7 @@ def create_enumeration(
         size = None
         for a in aspects:
             if a.f_identifier.text == "Size":
-                size = create_math_expression(a.f_value, filename, identifier.parent)
+                size = create_math_expression(a.f_value, filename)
             if a.f_identifier.text == "Always_Valid":
                 if a.f_value:
                     av_expr = create_bool_expression(a.f_value, filename)
```

Reply:
As discussed separately, I removed the package parameters (dd607b8fa67d1206df1c256c8ca9b7436d70a7b0) and most of the optional parameters (2de46911).